processors/pod_aggregator.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package processors
+
+import (
+	"fmt"
+
+	"k8s.io/heapster/core"
+)
+
+type PodAggregator struct {
+	MetricsToAggregate []string
+	LabelsToPopulate   []core.LabelDescriptor
+}
+
+func (this *PodAggregator) Process(batch *core.DataBatch) (*core.DataBatch, error) {
+	result := core.DataBatch{
+		Timestamp:  batch.Timestamp,
+		MetricSets: make(map[string]*core.MetricSet),
+	}
+
+	for key, metricSet := range batch.MetricSets {
+		result.MetricSets[key] = metricSet
+		if metricSetType, found := metricSet.Labels[core.LabelMetricSetType.Key]; found && metricSetType == core.MetricSetTypePodContainer {
+			// Aggregating containers
+			podName, found := metricSet.Labels[core.LabelPodName.Key]
+			ns, found2 := metricSet.Labels[core.LabelNamespaceName.Key]
+			if found && found2 {
+				podKey := core.PodKey(ns, podName)
+				pod, found := result.MetricSets[podKey]
+				if !found {
+					pod = this.podMetricSet(metricSet.Labels)
+					result.MetricSets[podKey] = pod
+				}
+
+				for _, metricName := range this.MetricsToAggregate {
+					metricValue, found := metricSet.MetricValues[metricName]
+					if !found {
+						continue
+					}
+					aggregatedValue, found := pod.MetricValues[metricName]
+					if found {
+						if aggregatedValue.ValueType != metricValue.ValueType {
+							return nil, fmt.Errorf("PodAggregator: type not supported in %s", metricName)
+						}
+
+						if aggregatedValue.ValueType == core.ValueInt64 {
+							aggregatedValue.IntValue += metricValue.IntValue
+						} else if aggregatedValue.ValueType == core.ValueFloat {
+							aggregatedValue.FloatValue += metricValue.FloatValue
+						} else {
+							return nil, fmt.Errorf("PodAggregator: type not supported in %s", metricName)
+						}
+					} else {
+						aggregatedValue = metricValue
+					}
+					pod.MetricValues[metricName] = aggregatedValue
+				}
+			} else {
+				return nil, fmt.Errorf("No namespace and/or pod info in container %s: %v", key, metricSet.Labels)
```

Comment:
I don't agree. We should have pod and namespace specified for every pod container. If there is not probably source implementation doesn't work as expected.